tracing-opentelemetry/src/metrics.rs

```diff
@@ -0,0 +1,331 @@
+/// is to add a key-value pair to your `tracing::Event` that contains a specific
+/// prefix. One of the following:
+/// - `MONOTONIC_COUNTER_` (non-negative numbers): Used when the counter should
+///   only ever increase
+/// - `COUNTER_`: Used when the counter can go up or down
+/// - `VALUE_`: Used for discrete data points (i.e., summing them does not make
+///   semantic sense)
+///
+/// Examples:
+/// ```
+/// # use tracing::info;
+/// info!(MONOTONIC_COUNTER_FOO = 1);
+/// info!(MONOTONIC_COUNTER_BAR = 1.1);
+///
+/// info!(COUNTER_BAZ = 1);
+/// info!(COUNTER_BAZ = -1);
+/// info!(COUNTER_XYZ = 1.1);
+///
+/// info!(VALUE_QUX = 1);
+/// info!(VALUE_ABC = -1);
+/// info!(VALUE_DEF = 1.1);
+/// ```
+///
+/// # Mixing data types
+///
+/// ## Floating-point numbers
+///
+/// Do not mix floating-point and non-floating-point numbers for the same
+/// metric. Instead, if you are ever going to use floating-point numbers for a
+/// metric, cast any other usages of that metric to `f64`.
+///
+/// Don't do this:
+/// ```
+/// # use tracing::info;
+/// info!(MONOTONIC_COUNTER_FOO = 1);
+/// info!(MONOTONIC_COUNTER_FOO = 1.0);
+/// ```
+///
+/// Do this:
+/// ```
+/// # use tracing::info;
+/// info!(MONOTONIC_COUNTER_FOO = 1_f64);
+/// info!(MONOTONIC_COUNTER_FOO = 1.1);
+/// ```
+///
+/// This is because all data published for a given metric name must be the same
+/// numeric type.
+///
+/// ## Integers
+///
+/// Positive and negative integers can be mixed freely. The instrumentation
+/// provided by `tracing` assumes that all integers are `i64` unless explicitly
+/// cast to something else. In the case that an integer *is* cast to `u64`, this
+/// subscriber will handle the conversion internally.
+///
+/// To illustrate this:
+/// ```
+/// # use tracing::info;
+/// // The subscriber receives an i64
+/// info!(COUNTER_BAZ = 1);
+///
+/// // The subscriber receives an i64
+/// info!(COUNTER_BAZ = -1);
+///
+/// // The subscriber receives a u64, but casts it to i64 internally
+/// info!(COUNTER_BAZ = 1_u64);
+///
+/// // The subscriber receives a u64, but cannot cast it to i64 because of
+/// // overflow. An error is printed to stderr, and the metric is dropped.
+/// info!(COUNTER_BAZ = (i64::MAX as u64) + 1)
+/// ```
+///
+/// # Under-the-hood
+///
+/// This subscriber holds a set of maps, each map corresponding to a type of
+/// metric supported by OpenTelemetry. These maps are populated lazily. The
+/// first time that a metric is emitted by the instrumentation, a `Metric`
+/// instance will be created and added to the corresponding map. This means that
+/// any time a metric is emitted by the instrumentation, one map lookup has to
+/// be performed.
+///
+/// In the future, this can be improved by associating each Metric instance to
+/// its callsite. However, per-callsite storage is not yet supported by tracing.
+pub struct OpenTelemetryMetricsSubscriber {
+    meter: Meter,
+    instruments: Arc<RwLock<Instruments>>,
```

Comment:
it shouldn't be necessary to place a single global `RwLock` around *all* the different metric types. updating a metric value or creating a new metric will only mutate one of the `HashMap`s in the `Instruments` struct at a time. however, by placing the whole thing inside of one big lock, this forces *all* metric updates to be globally synchronized, even for different types of metric. in the current implementation, updating a `u64` value recorder metric (for example) will prevent another thread from recording a `f64` up-down counter at the same time.

i think we should make the scope of the `RwLock`s much more granular.

Reply:
Definitely agree.